internal/controller/streams.go

```diff
@@ -197,105 +165,269 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+func connectAndStreamCiliumNetworkFlows(logger *zap.SugaredLogger, sm *streamManager) error {
+	ciliumCtx, ciliumCancel := context.WithCancel(context.Background())
+	defer ciliumCancel()
+
+	sendCiliumNetworkFlowsStream, err := sm.streamClient.client.SendKubernetesNetworkFlows(ciliumCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.ciliumNetworkFlowsStream = sendCiliumNetworkFlowsStream
+
+	err = sm.StreamCiliumNetworkFlows(ciliumCtx, sm.streamClient.ciliumNamespace)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream Cilium network flows", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamResources creates resourceStream client and begins the streaming of resources.
+func connectAndStreamResources(logger *zap.SugaredLogger, sm *streamManager) error {
+	resourceCtx, resourceCancel := context.WithCancel(context.Background())
+	defer resourceCancel()
+
+	SendKubernetesResourcesStream, err := sm.streamClient.client.SendKubernetesResources(resourceCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.resourceStream = SendKubernetesResourcesStream
+
+	err = sm.StreamResources(resourceCtx, resourceCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream resources", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamLogs creates sendLogs client and begins the streaming of logs.
+func connectAndStreamLogs(logger *zap.SugaredLogger, sm *streamManager) error {
+	logCtx, logCancel := context.WithCancel(context.Background())
+	defer logCancel()
+
+	SendLogsStream, err := sm.streamClient.client.SendLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.logStream = SendLogsStream
+	sm.bufferedGrpcSyncer.UpdateClient(sm.streamClient.logStream, sm.streamClient.conn)
+
+	err = sm.StreamLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream logs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Generic function to manage any stream with backoff and reconnection logic.
+func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
+	const (
+		initialBackoff       = 1 * time.Second
+		maxBackoff           = 1 * time.Minute
+		maxJitterPct         = 0.20
+		resetPeriod          = 10 * time.Minute
+		severeErrorThreshold = 10 // Define what constitutes a severe error.
+	)
+
+	var (
+		backoff             = initialBackoff
+		consecutiveFailures = 0
+		disableCilium       = make(chan bool)
+	)
 	max := big.NewInt(3)
+
+	resetTimer := time.NewTimer(resetPeriod)
 	for {
-		// Generate a random number
-		randomInt, err := rand.Int(rand.Reader, max)
-		if err != nil {
-			logger.Errorw("Could not generate a random int", "error", err)
+		select {
+		case <-resetTimer.C:
```

Comment:
This case will almost never be selected because of the `default` below.
In a `select`, if all the `case`s are blocking, the `default` is selected.